galley/pkg/kube/source/monitoring.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.opencensus.io/stats/view"
+	"context"
+)
+
+const (
+	apiVersion = "apiVersion"
+	group = "group"
+	kind = "kind"
+	errorStr = "error"
+)
+
+var (
+	APIVersionTag tag.Key
+	GroupTag      tag.Key
+	KindTag       tag.Key
+	ErrorTag tag.Key
+)
+
+var (
+	listenerHandleEventError = stats.Int64(
+		"galley/kube/source/listener_handle_event_error_total",
+		"The number of times the listener's handleEvent has errored",
+		stats.UnitDimensionless)
+
+	listenerHandleEventSuccess = stats.Int64(
+		"galley/kube/source/listener_handle_event_success_total",
+		"The number of times the listener's handleEvent has succeeded",
+		stats.UnitDimensionless)
+
+	sourceConversionSuccess = stats.Int64(
+		"galley/kube/source/source_converter_success_total",
+		"The number of times resource conversion in the source's ProcessEvent has succeeded",
+		stats.UnitDimensionless)
+	sourceConversionFailure = stats.Int64(
+		"galley/kube/source/source_converter_failure_total",
+		"The number of times resource conversion in the source's ProcessEvent has failed",
+		stats.UnitDimensionless)
+)
+
+func recordHandleEventError(msg string) {
+	ctx, ctxErr := tag.New(context.Background(), tag.Insert(ErrorTag, msg))
+	if ctxErr != nil {
+		scope.Errorf("error creating context to record handleEvent error")
+	} else {
+		stats.Record(ctx, listenerHandleEventError.M(1))
+	}
+}
+
+func recordHandleEventSuccess() {
+	stats.Record(context.Background(), listenerHandleEventSuccess.M(1))
+}
+
+type contextKey struct {
+	apiVersion, group, kind string
+}
+
+var ctxCache = make(map[contextKey]context.Context)
+
+func recordConverterResult(success bool, apiVersion, group, kind string) {
+	var metric *stats.Int64Measure
+	if success {
+		metric = sourceConversionSuccess
+	} else {
+		metric = sourceConversionFailure
+	}
+	key := contextKey{apiVersion, group, kind}
+	ctx, ok := ctxCache[key]
```

Comment:
@ozevren, can galley/pkg/kube/source::ProcessEvent be called from multiple goroutines? If so, this needs to be guarded with a lock. [`sync.Map`](https://golang.org/pkg/sync/#Map) would be appropriate. 

nit: `var err error` is only used within the `if !ok {}` block. Move it to line 89 immediately before `tag.New` to limit its scope.

Reply:
Yes, it is certainly multi-threaded.

Reply:
Ok, using sync.Map.